src/dbnode/encoding/m3tsz/m3tsz.go

```diff
@@ -61,6 +77,95 @@ var (
 	errInvalidMultiplier = errors.New("supplied multiplier is invalid")
 )
 
+// WriteTSZXOR writes the TSZ XOR into the provided stream given the previous
+// XOR and the current XOR.
+func WriteXOR(
+	stream encoding.OStream,
+	prevXOR, curXOR uint64) {
+	if curXOR == 0 {
+		stream.WriteBits(OpcodeZeroValueXOR, 1)
+		return
+	}
+
+	// NB(xichen): can be further optimized by keeping track of leading and trailing zeros in enc.
+	prevLeading, prevTrailing := encoding.LeadingAndTrailingZeros(prevXOR)
+	curLeading, curTrailing := encoding.LeadingAndTrailingZeros(curXOR)
+	if curLeading >= prevLeading && curTrailing >= prevTrailing {
+		stream.WriteBits(OpcodeContainedValueXOR, 2)
+		stream.WriteBits(curXOR>>uint(prevTrailing), 64-prevLeading-prevTrailing)
+		return
+	}
+
+	stream.WriteBits(OpcodeUncontainedValueXOR, 2)
+	stream.WriteBits(uint64(curLeading), 6)
+	numMeaningfulBits := 64 - curLeading - curTrailing
+	// numMeaningfulBits is at least 1, so we can subtract 1 from it and encode it in 6 bits
+	stream.WriteBits(uint64(numMeaningfulBits-1), 6)
+	stream.WriteBits(curXOR>>uint(curTrailing), numMeaningfulBits)
+}
+
+// WriteIntSig writes the number of significant bits of the diff if it has changed and
+// updates the IntSigBitsTracker.
+func WriteIntSig(os encoding.OStream, sigTracker *IntSigBitsTracker, sig uint8) {
```

Comment:
Why not make this a method on `IntSigBitsTracker`?  You could also set `encoding.OStream` on that too

Reply:
In fact what about creating a new struct called `OutputStream` in the M3TSZ package, then doing this:

```go


// now on Encoder you can embed it
type Encoder struct {
  stream OutputStream
}

func (e *Encoder) writeWhatever() {
  e.stream.os.Write(...)
}

func (e *Encoder) writeSomethingWithXOR() {
  e.stream.WriteXOR(prevXOR, curXOR)
}

func (e *Encoder) writeSomethingWithIntSig() {
  e.stream.WriteIntSig(sig)
}
```

Reply:
sure thats cleaner